scripts/utils/consensus-celltype-plotting-functions.R

```diff
@@ -0,0 +1,183 @@
+    dplyr::left_join(markers_df, by = "ensembl_gene_id", relationship = "many-to-many")
+
+  # prep for plots
+  # get total number of cells per final annotation group
+  total_cells_df <- consensus_df |>
+    dplyr::select(library_id, barcodes, broad_celltype_group) |>
+    dplyr::distinct() |>
+    dplyr::count(broad_celltype_group, name = "total_cells")
+
+  # table with one row per unique broad cell type/ marker gene combination
+  # first all cells in with the same broad_celltype_group (determined based on consensus_annotation) are grouped together
+  # then get the mean gene expression and total percentage of cells that express each marker gene across all cells in that group
+  # do this for every possible marker gene/ validation group assignment
+  # second we calculate the mean expression and mean percentage of all marker genes in a given validation group (this value is used only in the second section of the report)
+  group_stats_df <- consensus_df |>
+    # for each assigned cell type/marker gene combo get total detected and mean expression
+    # group by both broad group and validation group to account for genes that are expressed in more than one cell type
+    dplyr::summarize(
+      .by = c("broad_celltype_group", "ensembl_gene_id", "validation_group_annotation"),
+      detected_count = sum(detected),
+      mean_exp = mean(logcounts)
+    ) |>
+    # add in validation group for marker genes
+    # this includes all possible marker genes and all possible validation group assignments
+    dplyr::left_join(markers_df, by = c("ensembl_gene_id", "validation_group_annotation"), relationship = "many-to-many") |> # add total cells
+    dplyr::left_join(total_cells_df, by = c("broad_celltype_group")) |>
+    # for plotting we're only going to look at any cell types with > 50 cells otherwise these plots can get wild
+    dplyr::filter(total_cells > 50) |>
+    dplyr::mutate(
+      # get total percent expressed
+      percent_exp = (detected_count / total_cells) * 100,
+      # account for NA/unknowns and set axes order
+      broad_celltype_group = tidyr::replace_na(broad_celltype_group, "unknown") |>
+        factor(levels = c(unique(markers_df$validation_group_annotation), "unknown"))
+    )
+
+  # get list of celltypes to keep and assign colors
+  celltype_groups <- group_stats_df |>
+    dplyr::pull(broad_celltype_group) |>
+    unique() |>
+    as.character()
+
+  # filter markers to those that are actually relevant
+  # we will only plot the marker genes for cell types that are part of the assigned broad validation group for this group of samples
+  # we don't care about plotting marker genes for cell types that aren't present here
+  filtered_markers_df <- markers_df |>
+    dplyr::filter(
+      validation_group_annotation %in% celltype_groups,
+      gene_symbol %in% group_stats_df$gene_symbol
+    )
+
+  # specify x axis order for dotplot
+  marker_gene_order <- filtered_markers_df |>
+    dplyr::pull(gene_symbol)
```

Comment:
Everything seems to be working as expected here, but I am a bit worried by something I read that said that duckdb does not guarantee order of tables. With the current tables and plots, everything seems fine, but I wonder if we might want to make this more rigid/fixed somehow? 
I imagine you could fix these by using the order from `names(celltype_colors)`

But this absolutely might be a "not a problem now, address if it comes up" situation.

Reply:
I think the reason this isn't an issue here is because we are using regular `readr::read_tsv` to read in the marker genes table and save it. So I don't think it's using `duckplyr` here, but I can add something to make sure the order matches.